### PR TITLE
Dynamic route transitions

### PR DESCRIPTION
**Pull request by tutok**
### Please check if the PR fulfills these requirements ###

- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Changelog have been added / updated (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

Please fetch and play with it firstly

**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Improvement

**What is the current behavior? (You can also link to an open issue here)**
`<SwitchDynamicURL />` component extends standard `<Switch/>` (from `react-router`) and it is able to handle dynamic URLs. In order to render component for a resource type, you need to provider `type` (instead of `path`) in `<Route />` component, for example:

```
 <SwitchDynamicURL onLoading={({ component }) => <LoaderWrapper>{component}</LoaderWrapper>}>
  <Route exact path="/" component={Home} />
  <Route exact path="/blog/:page?" component={Blog} />
  <Route exact path="/cart" component={Cart} />
  <Route exact path="/checkout" component={Checkout} />
  <Route exact path="/checkout/confirmation" component={CheckoutConfirmation} />
  <Route exact path="/checkout/failure" component={CheckoutFailure} />
  <ProtectedRoute path="/account" component={Account} />
  <OnlyUnauthenticatedRoute exact path="/sign-in" component={SignIn} />
  <OnlyUnauthenticatedRoute exact path="/reset-password" component={ResetPassword} />
  <Route exact type="blog-post" component={BlogPost} />
//             ^
  <Route exact type="shop-category" component={Category} />
  <Route exact type="shop-product" component={Product} />
  <p>not Found</p>
</SwitchDynamicURL>
```

this allows you to define routing in a declarative way, so you have full control over what and how is rendered. Even when additional data needs to be fetched before.

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)**
NO

**Other information:**

This PR is about making possible to build transitions during navigation, and made possible to build pages according to `react-router` `v4` config less pattern (where we should be able to render many unrelated`Switch` components, like in `header, `main` or `footer`)

### CODE REVIEW

File: examples/shop-with-blog/client/src/App.js [hunk 1]
```diff
@@ -96,18 +100,23 @@ const App = () => (
...
+                            <Route exact path="/cart" component={Cart} />
+                            <Route exact path="/checkout" component={Checkout} />
+                            <Route exact path="/checkout/confirmation" component={CheckoutConfirmation} />
+                            <Route exact path="/checkout/failure" component={CheckoutFailure} />
+                            <ProtectedRoute path="/account" component={Account} />
+                            <OnlyUnauthenticatedRoute exact path="/sign-in" component={SignIn} />
+                            <OnlyUnauthenticatedRoute exact path="/reset-password" component={ResetPassword} />
+                            <Route exact type="blog-post" component={BlogPost} />
+                            <Route exact type="shop-category" component={Category} />
+                            <Route exact type="shop-product" component={Product} />
```
Comment (Borales) [inline examples/shop-with-blog/client/src/App.js]: 👍

Comment (tutok) [reply examples/shop-with-blog/client/src/App.js]: glad to hear, was worried that this will be next thinking to discuss :P

**Comment by Borales** [comment]
- product pages - nice! the second visit feels like a static page, instant load, no loaders and there's a background refetch query
- category pages - significantly faster, but loader is showing every time Category page is requested

**Comment by tutok** [comment]
thanks. Right now the Category page works in the same way as the Product page.

**Comment by tutok** [comment]
@Borales  I did not remove the previous `DynamicRouring` component. This PR does not introduce breaking changes - yet

**Comment by tutok** [comment]
@Borales @mateuszwozniak  should I remove old `DynamicRoute` component?

**Comment by Borales** [comment]
@tutok oh right, my mistake. Maybe we could keep it (temporary) for gradual migration

**Pull request merged**